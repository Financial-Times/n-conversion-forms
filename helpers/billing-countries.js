import { billingCountries } from 'n-common-static-data';

export { billingCountries };
