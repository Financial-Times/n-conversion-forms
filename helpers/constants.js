const printRegions = {
	cemeaV1: 'CEMEA_V1',
	cemeaV2: 'CEMEA_V2',
	apac: 'APAC',
};

const addressLine1AutoComplete = 'address-line1';
const addressLine2AutoComplete = 'address-line2';
const addressLine3AutoComplete = 'address-line3';

module.exports = {
	printRegions,
	addressLine1AutoComplete,
	addressLine2AutoComplete,
	addressLine3AutoComplete,
};
