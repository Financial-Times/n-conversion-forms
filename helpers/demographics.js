const { demographics } = require('n-common-static-data');

export { demographics };
