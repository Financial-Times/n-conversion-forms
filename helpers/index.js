module.exports = {
	'ncf-countries': require('./ncf-countries'),
	'ncf-common-data': require('./ncf-common-data'),
	cemeaV1ISO: require('./supportedCountries').cemeaV1ISO,
	cemeaV2ISO: require('./supportedCountries').cemeaV2ISO,
	apacISO: require('./supportedCountries').apacISO,
	identifyFTShippingZone: require('./supportedCountries')
		.identifyFTShippingZone,
	countriesSupported: require('./supportedCountries').countriesSupported,
	countriesSupportedISO: require('./supportedCountries').countriesSupportedISO,
	supportedPostcodeValidators: require('./supportedPostcodeValidators')
		.supportedPostcodeValidators,
	allSupportedPostcodeExamples: require('./supportedPostcodeExamples')
		.allSupportedPostcodeExamples,
	demographics: require('./demographics').demographics,
	billingCountries: require('./billing-countries').billingCountries,
};
