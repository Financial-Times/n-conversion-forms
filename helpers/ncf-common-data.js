const _get = require('lodash.get');
const COMMON_DATA = require('n-common-static-data');

module.exports = function ({ hash, fn }) {
	const context = Object.assign({}, { [hash.export]: _get(COMMON_DATA, hash.import) }, this);
	return fn(context);
};
