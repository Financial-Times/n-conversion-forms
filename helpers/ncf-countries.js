const { countries } = require('n-common-static-data').billingCountries;

const MINIMUM_TO_GROUP = 50;
const MINIMUM_TO_SHOW_GROUPS = 2;

module.exports = function ({ hash = {}, fn }) {
	const filter = Array.isArray(hash.filterList) ? hash.filterList : false;
	const value = hash.value;

	let data = countries;

	// Only show countries in the filter
	if (filter) {
		data = data.filter(item => filter.includes(item.code));
	}

	// Select the country
	if (value) {
		data = data.map(item => {
			item.selected = item.code === value;
			return item;
		});
	}

	// Group countries
	if (data.length >= MINIMUM_TO_GROUP) {
		data = groupCountries(data);
	}

	const context = Object.assign({ countries: data }, this);
	return fn(context);
};

/**
 * Produce a frequently used option group
 * @param {Array} countries
 * @return {Array}
 */
function groupCountries (countries) {
	const frequentlyUsed = ['GBR', 'USA', 'JPN', 'FRA', 'CAN'];

	const frequentlyUsedCountries = countries
		.filter(item => frequentlyUsed.includes(item.code))
		.map(item => Object.assign({}, item))
		.sort((a, b) => frequentlyUsed.indexOf(a.code) - frequentlyUsed.indexOf(b.code));

	// Not worth showing return standard countries list
	if (frequentlyUsedCountries.length < MINIMUM_TO_SHOW_GROUPS) {
		return countries;
	}

	// If selected country is in frequently used only leave that one selected
	// otherwise the user is shown the lowest selected option
	const selected = frequentlyUsedCountries.find(item => item.selected);
	if (selected) {
		countries.forEach(item => item.selected = false);
	}

	return [
		{ label: 'Frequently Used', countries: frequentlyUsedCountries },
		{ label: 'Alphabetical', countries }
	];
}
