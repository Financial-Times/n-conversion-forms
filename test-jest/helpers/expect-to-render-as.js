import ReactDOMServer from 'react-dom/server';
import toDiffableHtml from 'diffable-html';

export const expectToRenderAs = {
	toRenderAs(reactComponent, context, props) {
		// Render the React component with props.
		const ReactComponentRendered = ReactDOMServer.renderToStaticMarkup(
			reactComponent(props)
		);
		// Render with the same properties but using Handlebars.
		const handlerBarsVersion = context.template(props);
		// `this.isNot` indicates whether the assertion was inverted with `.not`
		// which needs to be respected, otherwise it fails incorrectly.
		if (this.isNot) {
			expect(toDiffableHtml(ReactComponentRendered)).notToBe(
				toDiffableHtml(handlerBarsVersion)
			);
		} else {
			// Compare the 2 rendered versions to be sure they match and compatibility is preserved.
			expect(toDiffableHtml(ReactComponentRendered)).toBe(
				toDiffableHtml(handlerBarsVersion)
			);
		}

		expect(ReactComponentRendered).toMatchSnapshot();
		expect(handlerBarsVersion).toMatchSnapshot();

		// This point is reached when the above assertions were successful.
		// The test should therefore always pass, that means it needs to be
		// `true` when used normally, and `false` when `.not` was used.
		return {
			pass: !this.isNot,
		};
	},
};
