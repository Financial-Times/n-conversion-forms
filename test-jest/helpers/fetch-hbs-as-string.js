const fs = require('fs');
const cheerio = require('cheerio');
const promisify = require('util').promisify;
const Handlebars = require('@financial-times/n-handlebars').handlebars;

const readFile = promisify(fs.readFile);
const PARTIAL_DIR = __dirname + '/../../partials/';

const handlebars = Handlebars();

const fetchPartial = async (name, returnString = false) => {
	let file = await readFile(PARTIAL_DIR + name, 'utf8');
	// HACK ALERT: this is necessary to make testing @partial-block work. It does mean that any test where
	//  a @partial-block helper isn't registered will blow up, but that will just have to be worked around
	//  by always registering it - even with an empty value if necessary.
	//  We need to use the `#if` around the partial block to make using that in a template optional.
	file = file.replace(/{{#if @partial-block}}([\s\S]*){{\/if}}/gm, '$1');
	const template = handlebars.compile(file);
	if (returnString) {
		return (context) => template(context);
	}
	return (context) => cheerio.load(template(context));
};

const fetchPartialAsString = async (name) => await fetchPartial(name, true);

module.exports = {
	fetchPartialAsString
};
