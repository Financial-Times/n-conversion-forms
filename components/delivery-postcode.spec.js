import { DeliveryPostcode } from './index';
import { expectToRenderAs } from '../test-jest/helpers/expect-to-render-as';
import { fetchPartialAsString } from '../test-jest/helpers/fetch-hbs-as-string';

const context = {
};

expect.extend(expectToRenderAs);

describe('Delivery Postcode', () => {
	beforeAll(async () => {
		context.template = await fetchPartialAsString('delivery-postcode.html');
	});

	it('render a postcode input with a label set as postcode', () => {
		const props = {
			postcodeReference: 'postcode',
			pattern: 'whatever',
		};

		expect(DeliveryPostcode).toRenderAs(context, props);

	});

	it('render a postcode input with a label set as Zip Code', () => {
		const props = {
			country: 'USA',
			postcodeReference: 'Zip Code',
			pattern: 'whatever',
		};

		expect(DeliveryPostcode).toRenderAs(context, props);

	});

	it('render a postcode input with a label set as Zip Code with USA in lower case', () => {
		const props = {
			country: 'usa',
			postcodeReference: 'Zip Code',
			pattern: 'whatever',
		};

		expect(DeliveryPostcode).toRenderAs(context, props);

	});

	it('render a postcode input with a label set as postal code', () => {
		const props = {
			country: 'CAN',
			postcodeReference: 'postal code',
			pattern: 'whatever',
		};

		expect(DeliveryPostcode).toRenderAs(context, props);

	});

	it('render a postcode input with default label', () => {
		const props = {
			postcodeReference: 'postcode',
			pattern: 'whatever',
		};

		expect(DeliveryPostcode).toRenderAs(context, props);

	});

	it('render a disable input', () => {
		const props = {
			postcodeReference: 'postcode',
			pattern: 'whatever',
			isDisabled: true,
		};

		expect(DeliveryPostcode).toRenderAs(context, props);

	});

	it('render different styles', () => {
		const props = {
			postcodeReference: 'postcode',
			pattern: 'whatever',
			hasError: true,
			isHidden: true,
		};

		expect(DeliveryPostcode).toRenderAs(context, props);
	});
});
