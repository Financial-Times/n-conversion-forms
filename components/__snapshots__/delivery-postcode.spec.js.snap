// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delivery Postcode render a disable input 1`] = `
<label id="deliveryPostcodeField"
       class="o-forms-field"
       data-validate="required"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery
      <span data-reference="postcode">
        ZipCode
      </span>
    </span>
  </span>
  <span class="o-forms-input o-forms-input--text">
    <input type="text"
           id="deliveryPostcode"
           name="deliveryPostcode"
           placeholder="Enter your ZipCode"
           autocomplete="postal-code"
           data-trackable="delivery-postcode"
           aria-required="true"
           required
           pattern="whatever"
           disabled
           value
    >
    <span class="o-forms-input__error">
      Please enter a valid
      <span data-reference="postcode">
        ZipCode
      </span>
      .
    </span>
  </span>
</label>
`;

exports[`Delivery Postcode render a disable input 2`] = `
<label id="deliveryPostcodeField"
       class="o-forms-field"
       data-validate="required"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery
      <span data-reference="postcode">
        ZipCode
      </span>
    </span>
  </span>
  <span class="o-forms-input o-forms-input--text">
    <input type="text"
           id="deliveryPostcode"
           name="deliveryPostcode"
           placeholder="Enter your ZipCode"
           autocomplete="postal-code"
           data-trackable="delivery-postcode"
           aria-required="true"
           required
           pattern="whatever"
           disabled
           value
    >
    <span class="o-forms-input__error">
      Please enter a valid
      <span data-reference="postcode">
        ZipCode
      </span>
      .
    </span>
  </span>
</label>
`;

exports[`Delivery Postcode render a postcode input with a label 1`] = `
<label id="deliveryPostcodeField"
       class="o-forms-field"
       data-validate="required"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery
      <span data-reference="postcode">
        ZipCode
      </span>
    </span>
  </span>
  <span class="o-forms-input o-forms-input--text">
    <input type="text"
           id="deliveryPostcode"
           name="deliveryPostcode"
           placeholder="Enter your ZipCode"
           autocomplete="postal-code"
           data-trackable="delivery-postcode"
           aria-required="true"
           required
           pattern="whatever"
           value
    >
    <span class="o-forms-input__error">
      Please enter a valid
      <span data-reference="postcode">
        ZipCode
      </span>
      .
    </span>
  </span>
</label>
`;

exports[`Delivery Postcode render a postcode input with a label 2`] = `
<label id="deliveryPostcodeField"
       class="o-forms-field"
       data-validate="required"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery
      <span data-reference="postcode">
        ZipCode
      </span>
    </span>
  </span>
  <span class="o-forms-input o-forms-input--text">
    <input type="text"
           id="deliveryPostcode"
           name="deliveryPostcode"
           placeholder="Enter your ZipCode"
           autocomplete="postal-code"
           data-trackable="delivery-postcode"
           aria-required="true"
           required
           pattern="whatever"
           value
    >
    <span class="o-forms-input__error">
      Please enter a valid
      <span data-reference="postcode">
        ZipCode
      </span>
      .
    </span>
  </span>
</label>
`;

exports[`Delivery Postcode render different styles 1`] = `
<label id="deliveryPostcodeField"
       class="o-forms-field ncf__hidden"
       data-validate="required"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery
      <span data-reference="postcode">
        ZipCode
      </span>
    </span>
  </span>
  <span class="o-forms-input o-forms-input--text o-forms-input--invalid">
    <input type="text"
           id="deliveryPostcode"
           name="deliveryPostcode"
           placeholder="Enter your ZipCode"
           autocomplete="postal-code"
           data-trackable="delivery-postcode"
           aria-required="true"
           required
           pattern="whatever"
           value
    >
    <span class="o-forms-input__error">
      Please enter a valid
      <span data-reference="postcode">
        ZipCode
      </span>
      .
    </span>
  </span>
</label>
`;

exports[`Delivery Postcode render different styles 2`] = `
<label id="deliveryPostcodeField"
       class="o-forms-field ncf__hidden"
       data-validate="required"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery
      <span data-reference="postcode">
        ZipCode
      </span>
    </span>
  </span>
  <span class="o-forms-input o-forms-input--text o-forms-input--invalid">
    <input type="text"
           id="deliveryPostcode"
           name="deliveryPostcode"
           placeholder="Enter your ZipCode"
           autocomplete="postal-code"
           data-trackable="delivery-postcode"
           aria-required="true"
           required
           pattern="whatever"
           value
    >
    <span class="o-forms-input__error">
      Please enter a valid
      <span data-reference="postcode">
        ZipCode
      </span>
      .
    </span>
  </span>
</label>
`;
